### DISCUSSION TITLE
Demo repository

### DISCUSSION THREAD

**Discussion by amoeamoe** [nicolaasburgers/definition-library-releases#5] [General]
Could you create a public demo repository with some sample components that anyone can connect to in order to see the structure in a repository with the correct format? I'm unfortunately not able to add any components without crashing Rhino. I'm able to authenticate, but adding a URL to the Saved Connections list does not seem to clone any repository to the selected destination folder. Pressing the load button does nothing and the other buttons are always grey, even when a cluster is selected. Any suggestions for what I should check?

**Comment by nicolaasburgers** [comment]
Hi thanks for that question - I'm clearly early on in my documentation effort (at time of writing this) and you've pointed out something I really need to clarify:

This plug-in does not clone a repository or create any local copy of a whole repository.  It communicates with GitHub like a remote document database, through an API.  This means you don't need to worry about any file/folder/repo structure - the plug-in will just take your files/clusters and communicate with GitHub as required to update your repository.  Note: the plug-in assumes you won't alter the repository manually yourself, so just leave the repo to just be modifed by the plug-in only.

The default local folder for files is used by the plug-in to store files downloaded from the GitHub servers, but that's not the same as cloning.  The way I've used the Revision entries in the document's properties, you can move the files to any location and the plug-in will be able to know that its library name, version, and which repository it was downloaded from - which wouldn't be the case if I used local cloned repositories.

So if you have a private repository in your GitHub account already (it can just be blank), if can authenticate, then after saving the GH file locally (which I've currently got as a prerequisite for publishing) you can go to Library -> Publish, enter the name and tags, etc and press Publish.

If it's a blank repository, it will do what it needs to do to the repository (including a simple README file if the repo is blank) before uploading it using the GitHub API.

I'm curious about your crashing though and what you mean by adding components - can you elaborate with the steps you take and any useful screenshots you can capture?  Are you using version 0.1.2?

By the way, there might be logs that you could send - if you go to Library -> Settings then "Go to folder" next to the logs directory, you might find some .txt log files there that might be of use.

**Reply by amoeamoe** [reply]
Oh, that is very cool, thank you for elaborating. Perhaps I didn't save the file first, I'll look into it and send you the logs if I'm unable to make it work. When I was talking about adding components I meant sharing clusters. Very impressed by the functionality of your tool by the way, it seems especially useful for projects where we work with other firms and can't use our in house versioning system. I would recommend that you make it available from the package manager, then we can install it automatically as a part of the default Rhino setup within our firm.

**Reply by amoeamoe** [reply]
I tested the plugin again and now it works as expected. The issue I experience might be related to another plugin I have that also edits the document properties, but once I disabled it Rhino never crashed again.

**Reply by nicolaasburgers** [reply]
It's great to hear that it now works as expected!  In terms of package manager, I do indeed intend to make it available there from the beta version onwards.

You mention a few very interesting things there:
- can you talk a bit about the in-house versioning system you use?
- can you name the other plug-in you use which edits the document properties?  If it's publicly avaliable I'd like to do my own testing with that plug-in too to see if I can make them more compatible

**Reply by amoeamoe** [reply]
The in-versioning system is still under development, but currently it exist in three different formats. 
1. An in-house server for the package manager to distribute things compiled as yak packages.
2. A small python script that runs when Rhino is started that collects the latest version of any user object from a network drive. These objects are shared with a custom menu button.
3. A browse for script tool to access best practice Grasshopper scripts. The Grasshopper scripts are shared on the in-house git platform COWI Git and the script metadata is stored in the document properties as you did. This plugin use a Grasshopper template I think might have caused the compatibility issue, but I'm not sure. I can't send you the plugin as this comment does not support .gh files, but I've attached what you need to create the template. Take a python node and change it to SDK mode. Create a cluster containing the python node and paste the attached code inside the python node. It should give you the template setup.
[SharedTemplate.txt](https://github.com/nicolaasburgers/definition-library-releases/files/13764847/SharedTemplate.txt)

I have some other question for you actually. To upgrade your clusters, do you use the IGH_UpgradeObject method? And the rendering adjustments of the clusters with name and versions behind, is that drawn directly on the canvas or is it a update of the cluster attributes or something completely different? I'm surprised it does not slow down the redraw speed of the canvas,

**Reply by nicolaasburgers** [reply]
Thanks I'll have a go at recreating the crashing using the instructions in your point 3.  If you can't share `.gh` files on this platform then feel free to email them (in a zip or plain) to `work.nicolaas@gmail.com`.

To be extra precise about what I store in the document properties: I store a base64-encoded JSON string containing the published name, version, repo URL and relative path within that repo.  If you want to have a look you can just decode (base64) that string yourself.  I chose this simple encoding just to discourage manual alteration rather than being an impenetrable method of storage.

To upgrade the clusters I use the GH_Cluster.UpdateDocument method.  I found the GH_Cluster type isn't well documented and this was one of several points in the writing of the code that came about through exploration/experimentation.

Another such point was the drawing of the name, version and version diff info above clusters.  I originally was thinking of creating my own cluster component to display that info (within the drawing surface of the component itself) but after discovering the Grasshopper.Instances.ActiveCanvas.CanvasPaintEnd event, I found to my surprise too that it was extremely fast at drawing the text and colours directly onto the canvas and I thought it looked cooler anyway; since it doesn't alter the definition itself, it ensures that it'll work fine for people who don't have the plug-in, and for people who do, it displays something very cool (but hopefully not intrusive) above each cluster.

**Reply by nicolaasburgers** [reply]
I haven't managed to reproduce a crash, but after going through the procedure you outlined I did reproduce an errored cluster where the meta data info would be rendered on the canvas by your attached python code.  

The python code assumes it can find COWI revision info at the last entry in the list, so I've made trial changes to the DefinitionLibrary code so that it inserts its own revision entries at the start of the list, preserving other existing revision entries at the end of the list, in their existing order.  This seems to avoid the error in the cluster, although it does cause revision entries to be out of DateTag order in the document.  If you can email me an example `.gh` file I can do another test.

I'm thinking if there are any other plug-ins out there that make assumptions about revision entries, they might be likely to gravitate towards the end of the list, so perhaps having my plug-in work at the start of the list might be both feasible and slightly defensive.

Other readers: if anyone has any thoughts on this concept, it would be greatly appreciated.
